pkg/controller/autoscalinggroup.go

```
@@ -0,0 +1,379 @@
+// return max.
+func getDesiredNodeCount(curr, min, max int) int {
+	if curr < min {
+		return min
+	} else if curr > max {
+		return max
+	}
+
+	return curr
+}
+
+func isScaleUpEvent(curr, desired int) bool {
+	return curr < desired
+}
+
+// syncHandler surveys the state of the systems, checking to see that the current
+// number of nodes selected by an autoscaling groups node selector is within the
+// autoscaling groups bounds for min and max
+func (agc *AutoscalingGroupController) syncHandler(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	autoScalingGroup, err := agc.agLister.Get(name)
+	if err != nil {
+		if kubeerrors.IsNotFound(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	if autoScalingGroup.Spec.Suspend {
+		log.Infof("Autoscaling Group '%s' was queued but it is currently suspended.", autoScalingGroup.Name)
+		return nil
+	}
+
+	engineName := autoScalingGroup.Spec.Engine
+	if !agc.autoscalingengine.IsAutoscalingEngine(engineName) {
+		log.Errorf("The Autoscaling Engine Specified for the Autoscaling Group '%s' is not registered.", autoScalingGroup.Name)
+		return nil
+	}
+
+	ns := getNodesLabelSelector(autoScalingGroup.Spec.NodeSelector)
+	// get nodes associated with autoscaling group using the node selector
+	nodes, _ := agc.nodeLister.List(ns)
+	numNodes := len(nodes)
+	log.Info("Current number of nodes in autoscaling group ", autoScalingGroup.Name, ": ", numNodes)
+
+	desired := getDesiredNodeCount(numNodes, autoScalingGroup.Spec.MinNodes, autoScalingGroup.Spec.MaxNodes)
+	// If number of nodes is within threshold then nothing needs to change this
+	// and this can noop and should just return
+	if numNodes == desired {
+		return nil
+	}
+
+	engine, err := agc.autoscalingengine.GetAutoscalingEngine(engineName)
+	if err != nil {
+		return err
+	}
+
+	scaled, err := engine.SetTargetNodeCount(ns, desired, "")
```

The heuristic passed should depend on whether it's a scale up or scale down. Along those lines, the heuristic would normally come from an AutoscalingPolicy, but this scale event isn't from a metric alert and thus can't be associated with a specific Policy. What's the expected behavior in this case? 🤔

---

I think it would just default to random? should there also be a heuristic on an ASG?

---

I think the AutoscalingGroup will need the heuristics regardless, as an AutoscalingPolicy should never know the implementation details of an AutoscalingEngine. I think this was overlooked in the RFC. @mattkelly is correct though that we will need to pass the proper heuristic here (up vs down). Defining it in the AutoscalingGroup does give us the ability to pass the proper heuristic here as you're alluding to @ashleyschuett

---

How should this look since the AutoscalingPolicy specifies a heuristic for both `ScaleUp` and `ScaleDown`. Should these properties also be added to the ASG? 

```
scaleup: 
 heuristic: random
scaledown:
  heuristic: cpu
```